### DISCUSSION TITLE
GT4286Util v0.0.0.1

### DISCUSSION THREAD

**Discussion by nk64** [nk64/GT4286#8] [Announcements]
First Release of GT4286Util.

Features:
- identify the generation of an SD Card
- refresh the game database from built-in and downloaded roms (you can add roms to the builtin /rom folder if you like)

<hr /><em>This discussion was created from the release <a href='https://github.com/nk64/GT4286/releases/tag/v0.0.0.1'>GT4286Util v0.0.0.1</a>.</em>

**Comment by nk64** [comment]
```
USAGE:
    GT4286Util [OPTIONS] <COMMAND>

EXAMPLES:
    GT4286Util identify D:
    GT4286Util identify C:\RetroControllerCardBackups\SDCARD1
    GT4286Util identify /mnt/sdcard
    GT4286Util refreshgamedb D:
    GT4286Util refreshgamedb C:\RetroControllerCardBackups\SDCARD1

OPTIONS:
    -h, --help       Prints help information
    -v, --version    Prints version information

COMMANDS:
    identify <pathtosdcard>         Attempt to identify the generation of the SD Card (or backup folder)
    refreshgamedb <pathtosdcard>    Update the game.db with new and removed roms in /roms and /download
```

**Reply by horsiezoo** [reply]
Thanks for this, works perfectly, obviously I’ve made so many experimental changes to various files it doesn’t recognise my sd as a known build but the db rebuild is great!

**Reply by nk64** [reply]
Could you try running it against your SD Card backup folder? you did take a backup right? :-)
If it is still not detected, please paste the details in here.

**Reply by horsiezoo** [reply]
Yep multiple bkups, pretty much first thing I did was copy supplied sd card to a faster and more reliable card

Dump : (its 157v) 


Unknown SD Card Generation. If this is a stock SD Card, please join the discussions at
https://github.com/nk64/GT4286/discussions and post these details

{
  "GenerationId": "unknown",
  "ApproximateDateDetails": "unknown",
  "GameListCount": 157,
  "GameDbBuiltinCount": 157,
  "TotalRomCount": 157,
  "RomDirFileCounts": {
    "/roms/FBA": 67,
    "/roms/FC": 69,
    "/roms/GB": 0,
    "/roms/GBA": 9,
    "/roms/GBC": 0,
    "/roms/MAME": 0,
    "/roms/MD": 6,
    "/roms/PCE": 0,
    "/roms/PS": 3,
    "/roms/SFC": 3
  },
  "KeyFileHashes": {
    "ui/gamelist.txt": "35A5E1019E9913FF6C4D3C7865D20159DD84A3D6FDA1409418DE59816D2F807C",
    "bin/gmenu2x": "424D75020480E23E56080FA5AD893E20D766A7AC49C4E16FDA70B06AB21B94AC",
    "emus/fbneo/fbneo": "BB9DA0976667A3656EB851582B3D02D49253983E8A10FCBDBEAA47EB25F74282",
    "emus/fceux/fceux": "809CA2B23E049D55F211099CF2AF2A0FF4A01240B023C245D455874A20FAD862",
    "emus/gamebatte/gambatte_sdl": "1CE888678AC6B5EB71081F09B60DF0A25A6196ADC32CE3733233522073CAA8FC",
    "emus/gbc/gambatte_sdl": "1CE888678AC6B5EB71081F09B60DF0A25A6196ADC32CE3733233522073CAA8FC",
    "emus/gpsp/gpsp": "AD6942C9B5CEAC8C401F13B624FA52CD5E297C8C72629647730C136BFF28BB31",
    "emus/mame/fbneo": "24B4026764C6B8F218D56222A534EDB3D21210A2C8FCDEE8A5F96505D6B85E9F",
    "emus/pcsx4all/pcsx": "56A02A9FCE4C2A6FB58C098FEA424078E208C4DEA7854B8B14246936FA2FAB6C",
    "emus/picodrive/PicoDrive": "AAADDB5934E6FDD0D64EF474E88F96C2D1E08B154DD5300BCD9CD9D6837A3AEE",
    "emus/snes9x4d/snes9x4d.dge": "4E7FA039B9D0950BD6D01571E4147EAACB9741D36D71C59B9E41DB28E28B38BD",
    "emus/temper/temper": "C829A70EFD7AD6E30F64B36CA20E0AB783DBCB274F1217FC7B758CE88AE1D68C"
  },
  "Title": "unknown",
  "Notes": "unknown"
}

**Reply by nk64** [reply]
It looks like that is a Gen 2024-05-26 but with a patched ```/emus/fbneo/fbneo``` file.

I should figure out how to account/allow for patched versions of the emulators.

**Comment by ot644** [comment]
Thanks for the great work being done on this so far. Will refreshing the game.db resolve the issue of launching games added to collections or from searches?

**Reply by nk64** [reply]
I haven't tried all the games, but I haven't had any issues searching and adding to the Collection/Favourites.

Can you be more specific about how to reproduce the issue you are having? also which generation of SD card you have. Or a link to any existing discussions about it?. Cheers.

**Reply by ot644** [reply]
The sd card I have is one of the later ones which contain the 157 games. From my experience I have no issue adding games to collections or searching for games. But when I try to launch them (from the collections tab or from the search results) they fail to launch. When I browse to them however from the downloads tab they work without issue. I haven’t tested yet with games from the original roms location so will be interesting to see if moving everything to the roms directory and refreshing the game.db will resolve.

**Reply by nk64** [reply]
I have tested this and I can see the problem you have described.  I have added it to the [Known Issues](https://github.com/nk64/GT4286/blob/main/Known%20Issues.md) page.

**Comment by ot644** [comment]
Also, Will moving all roms to the built in roms folder remove the functionality of allowing deletion of roms using the UI? From what I can see, deletions are only possible from downloads.

**Reply by nk64** [reply]
Yes. The main advantage of using the builtin roms folder is that it fixes this issue with certain emulators and the saved state files for downloaded games causing an extra non working entry to appear in that list.

From the [main page](https://github.com/nk64/GT4286)
> If you add roms to the appropriate emulator sub dir in /downloads, duplicate entries for some games might appear after saving a game and rebooting the console. This is because while most emulators save their game state in the emulator folder, some emulators eg GB, GBC, MD save their state in the rom folder and the next time the console is rebooted it will rescan the download folder and add a non working entry to the list. You will end up with a non working entry for every save state slot you use. (we can fix this).

**Reply by nk64** [reply]
Also as ot644 has described, downloaded games don't seem to work if you come back to them via the History or Collection tabs

**Comment by ot644** [comment]
Unfortunately when I attempt to run the utility executable Windows defender detects the below and deletes the file.

Trojan:Win32/Wacatac.B!ml

Alert level: Severe
Status: Active
Date: 29/12/2024 5:07 PM
Category: Trojan
Details: This program is dangerous and executes commands from an attacker.

**Reply by nk64** [reply]
Hmm, I imagine that would be unsettling and annoying for you.
I just uploaded it to VirusTotal and 1 out of 72 said 'Malicious'.

You can consider adding an exception in Defender for it or alternatively (and it might be a bridge too far) but you could try building the utility from source. I intend to put up some instructions for that but basically you need to:
1. install the [dotnet v8 sdk](https://dotnet.microsoft.com/en-us/download/dotnet/8.0)
2. clone this repo locally using git or download the current [contents of the repo as a zip file](https://github.com/nk64/GT4286/archive/refs/heads/main.zip) and unzip it somewhere.
3. then you can open a command prompt in the src\GT4286Util folder
4. and type ```dotnet publish```
5. an executable should end up in ```src\GT4286Util\bin\Release\net8.0\win-x64\publish```
6. try it and let me know what happens.

**Reply by ot644** [reply]
I simply added the file to the windows defender exclusions. The utility is brilliant. I cleaned up the original roms folder and moved all my download roms for a much cleaner experience. I can confirm that now any games I launch from favourites or launch using the search work without any issues. Also as you mentioned, files can no longer be deleted using the UI from this location which is also a good thing to protect from accidental deletions especially with young kids. Thanks again for all the work performed on this.

**Reply by nk64** [reply]
That's great, thanks for the feedback.

**Comment by benjiwest69page** [comment]
Does anyone know what to do if i have wiped the sd card? anywhere do download an image or firmware? thanks.

**Reply by horsiezoo** [reply]
Hi there, there may still be a dump on archive.org, hold fire I’ll check a bit later. Else could upload mine (It’s in storage somewhere)

**Reply by benjiwest69page** [reply]
The ones or archive are not complete. That would be amazing thank you so much.... I'm not very good and I was trying things that are over my head...

**Reply by horsiezoo** [reply]
https://archive.org/details/gt-4286-1619-games I think that’s it

ah just saw yr comment right I’ll dig mine out later

**Reply by horsiezoo** [reply]
Hmm just checked against the stock sd card and the above is exactly the same, format a 32gb card to FAT32 and extract the zip to that. 
When up and running you can swap the games out via NK64’s method (I have a heavily modified 64gb card full of random experiments but unfortunately that would take a very long time to upload)
The root of SD card should look like this : 

<img width="1453" height="987" alt="image" src="https://github.com/user-attachments/assets/813280ed-ebd8-479c-aa84-062c1e404a1c" />

**Reply by benjiwest69page** [reply]
Thank you.

**Reply by horsiezoo** [reply]
Hopefully any of that helps, it’s a wierd console, if anyone can figure out how to get it to map a usb keyboard I’d be very happy

**Reply by benjiwest69page** [reply]
Cherry pi is easier ☺️
